### ISSUE TITLE
🐛 `double free or corruption (!prev)` in a particular console

### ISSUE THREAD

**Issue by hololeap** [dandavison/delta#1288]
This seems to only be reproducible in a certain console window, which is integrated into [`kate`](https://kate-editor.org/) and is also running [`tmux`](https://github.com/tmux/tmux/). So far, I can't reproduce it in other console windows even if I adjust their size to be as small as the one displaying the bug.

I have `less` as my default pager and I have this line in my `~/.gitconfig`:

```gitconfig
	pager = delta --dark --line-numbers --side-by-side --syntax-theme 'Monokai Extended' --hyperlinks
```

To reproduce in this particular console, I:

1. Run `git diff --staged` (output has been copied [here](https://gist.github.com/hololeap/23eeccaf7f75d95809bf5ed9b472f54b))
2. Hit `<END>` to scroll to the bottom

`delta` crashes with the text `double free or corruption (!prev)`, and afterwards I have to run `reset` to get echo working in the terminal again. `tmux` doesn't seem to be affected otherwise.

Screenshot:

![image](https://user-images.githubusercontent.com/5141476/213931928-4b596175-b190-44c3-be1a-ceecd2814944.png)

**Comment by hololeap** [comment]
I have confirmed this is happening in this particular console window even when `tmux` isn't running.

I found that resizing the console to `156x48` prevents this from happening, but it still happens with e.g. `156x47`.

**Comment by SuperTux88** [comment]
I also just got this error. I'm using kitty as terminal, no tmux involved (no idea if that's relevant). And it doesn't even open for me, it just crashes directly with this message.

It only happens for some commands, for example I just did a `git diff <path to a file>` and it crashed, but if I do just a `git diff`, it doesn't crash. And running it with another file as parameter also works, so I have no idea what causes the crash.

But you are right, it seem to be related to terminal size, as if I resize the terminal (one less than full width), it also doesn't crash with the file that crashed before :man_shrugging: 

Any idea on how to debug this?

**Comment by SuperTux88** [comment]
Also, it's very weird, I can't reproduce the crash on another computer, with the same terminal size and the same diff of the same file and same delta version, it doesn't crash.

And I noticed on the computer where it does crash, I have multiple widths where it crashes, so 319-324 crash, <=318 and >= 325 doesn't crash (OK, I didn't test every single size, above or below that range, only a few).

**Comment by SuperTux88** [comment]
OK, this just got even weirder ... I just opened a new terminal window to try if I can still reproduce that, and it worked perfectly in that new window. But I still have the old terminal window open too (but had something else running there, but I then killed that to test there too), and it still crashes there. WTF? I was really confused.

I then tried to compare what's different between these windows, and I noticed, that the working directory of the new window was still my home directory (I stored the diff that causes the crash in a file, so I could do `delta < path/to/diff`, but that also meant, that it also works if I'm not in the git repo anymore) ... and sure, if I cd to the repo from where the diff was made, it crashes again there. So it looks like this crash not only depends on the terminal size but also on the working directory? :open_mouth:

So I then thought what's maybe different when I'm not in the actual repo. I have `hyperlinks = true` in my config, and they are still created if I'm not in the git repo as working directory, but they obviously don't work, as the files don't exist there. But just to make sure, I temporarily disabled hyperlinks, and the crash didn't happen anymore, enable it again, and it crashes again.

But what still confuses me, that the hyperlinks also get created when I'm not in the repo (they just don't work when clicking on it, because the file doesn't exist), but it doesn't crash then. But when I'm in the repo, but rename the file (so it also doesn't exist), it still crashes. I also tried to cd into another repo, but it also doesn't crash there. So it only crashes when hyperlinks are enabled and if in the same repo as the diff was created, unrelated to if the actual file that changed in the diff still exists in the working directly. But also somehow it depends on the terminal size?

Does anybody have an idea how all this is connected together? :thinking:

**Comment by dandavison** [comment]
> delta crashes with the text `double free or corruption (!prev)`, 

Are you sure it's delta that's crashing? Delta is written in Rust whereas I believe this error is associated with software written in C (or C++.). So that raises the question of whether it's less or tmux that's generating this error. Do you still get the error when delta is not invoking less? (e.g. `--paging=never`).

**Comment by SuperTux88** [comment]
Yes, that also confused me, but I thought that maybe same dependency of delta was linked against some C code which then was crashing. I kinda forgot that delta just calls less for the paging and that it could just be that, that's crashing.

In my setup, there is no tmux involved. But yes, if I call delta with `--paging=never` it doesn't crash, so it looks like it's less which is crashing. But then it's still weird that delta is generating something that causes less to crash in some cases? :thinking:

**Comment by dandavison** [comment]
Hin @SuperTux88, are you using a recent `less` version?

**Comment by SuperTux88** [comment]
I'm using `less 608 (PCRE2 regular expressions)` (from `less --version`), which looks like is the newest stable version from http://www.greenwoodsoftware.com/less/. There are `618_beta` and `623_beta` available in Gentoo, but they are masked, so not sure if I should give them a try (and they aren't listed on the website? :thinking:). But maybe I'll check if I can still reproduce the crash with these versions later.

**Comment by hololeap** [comment]
I am also using `less 608 (PCRE2 regular expressions)` in Gentoo.

**Comment by dandavison** [comment]
Maybe capture delta's output on disk and feed it to `less` separately, and reduce/bisect it to find the part that crashes less? That could make a good `less` bug report.

**Comment by hololeap** [comment]
@dandavison I think you are correct that this is an issue with `less`. I've reported it there.

**Comment by SuperTux88** [comment]
And the crash is fixed in `less` with this commit for me: https://github.com/gwsw/less/commit/c4bd5f2aa3d0e26cda0cc859804d48549b5d9156 :partying_face:

So I think this issue can also be closed, as it's not a bug in delta.
